Add support for opacity to the CSS parser fast-path

smfr opened the pull request:
#### 4d43216e1ba385baa7009dec90791314f7eb7e40
<pre>
Add support for opacity to the CSS parser fast-path
<a href="https://bugs.webkit.org/show_bug.cgi?id=276646">https://bugs.webkit.org/show_bug.cgi?id=276646</a>
<a href="https://rdar.apple.com/131808782">rdar://131808782</a>

Reviewed by Sam Weinig.

Add support for parsing opacity to CSSParserFastPaths, supporting numeric and percentage
forms.

Rearrange CSSParserFastPaths::maybeParseValue() to switch on the propertyID; previously we
we calling `parseSimpleTransform()` for things we knew were not a transform.

Hoist the call to `isSimpleLengthPropertyID()` out of `parseSimpleLengthValue()`. The
`isCSSViewportParsingEnabledForMode()` check there is obsolete and can probably be removed,
since we don&apos;t support `@viewport`. For simple length properties, we still fall through to
`parseKeywordValue()` for values like `auto`.

Support the `none` display value.

Remove `transformCanLikelyUseFastPath()` because it&apos;s faster to just call call
`parseSimpleTransformList()`.

* LayoutTests/fast/css/parsing-opacity-expected.txt:
* LayoutTests/fast/css/parsing-opacity.html:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isSimpleLengthPropertyID):
(WebCore::parseSimpleNumberOrPercentage):
(WebCore::parseSimpleLengthValue):
(WebCore::parseColorIntOrPercentage):
(WebCore::parseHexColorInternal):
(WebCore::parseNumericColor):
(WebCore::parseSimpleTransformList):
(WebCore::parseSimpleTransform):
(WebCore::parseDisplay):
(WebCore::parseOpacity):
(WebCore::CSSParserFastPaths::maybeParseValue):
(WebCore::transformCanLikelyUseFastPath): Deleted.
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):

Canonical link: <a href="https://commits.webkit.org/281010@main">https://commits.webkit.org/281010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d603e11c116124537f4218c25c50399c16ba70f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37721 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10878 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8838 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60522 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9035 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47294 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6302 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60424 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50496 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28137 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/32101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7795 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7842 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63723 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2307 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2315 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50512 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54678 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12896 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1950 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/33550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/34636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34381 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -940,6 +907,22 @@ static RefPtr<CSSValue> parseDisplay(StringView string)
+static RefPtr<CSSValue> parseOpacity(StringView string)
+{
+    double number;
+    CSSUnitType unit = CSSUnitType::CSS_NUMBER;
```
```suggestion
    auto unit = CSSUnitType::CSS_NUMBER;
```

nt1m commented on Source/WebCore/css/typedom/CSSUnitValue.cpp:
```diff
@@ -175,8 +175,8 @@ RefPtr<CSSValue> CSSUnitValue::toCSSValue() const
 {
-    bool acceptsNegativeNumbers = true;
-    if (CSSParserFastPaths::isSimpleLengthPropertyID(propertyID, acceptsNegativeNumbers) && !acceptsNegativeNumbers && value < 0)
+    ValueRange valueRange = ValueRange::All;
```
```suggestion
    auto valueRange = ValueRange::All;
```

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -950,17 +933,39 @@ static RefPtr<CSSValue> parseColorWithAuto(StringView string, const CSSParserCon
-        return result;
-    return parseSimpleTransform(propertyID, string);
+
+    ValueRange valueRange = ValueRange::All;
```
```suggestion
    auto valueRange = ValueRange::All;
```

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -150,14 +150,32 @@ static inline bool parseSimpleAngle(std::span<const CharacterType> characters, C
+        return false;
+
+    number = *parsedNumber;
+    if (number < 0  && valueRange == ValueRange::NonNegative)
```
```suggestion
    if (number < 0 && valueRange == ValueRange::NonNegative)
```

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -950,17 +933,39 @@ static RefPtr<CSSValue> parseColorWithAuto(StringView string, const CSSParserCon
+        ASSERT(CSSProperty::isColorProperty(propertyID));
+        return parseColor(string, context);
+    default:
+        break;
```
```suggestion
    default:
        if (CSSProperty::isColorProperty(propertyID))
            return parseColor(string, context);
        break;
```

Isn't this the correct code?

nt1m replied:
Alternatively move the if check outside of the switch.

smfr replied:
The patch isn't wrong, but duplicates a couple of lines.

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -950,17 +933,39 @@ static RefPtr<CSSValue> parseColorWithAuto(StringView string, const CSSParserCon
         return parseDisplay(string);
-    if ((propertyID == CSSPropertyCaretColor || propertyID == CSSPropertyAccentColor) && isExposed(propertyID, &context.propertySettings))
-        return parseColorWithAuto(string, context);
+    case CSSPropertyOpacity:
```
There are other properties that could benefit: fill-opacity/stroke-opacity/flood-opacity/stop-opacity (see `<alpha-value>` in CSSProperties.json)

smfr replied:
Those can be done separately.

nt1m reviewed (changes requested):
r- for the color change

weinig commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -47,7 +47,7 @@
 namespace WebCore {
 
-bool CSSParserFastPaths::isSimpleLengthPropertyID(CSSPropertyID propertyId, bool& acceptsNegativeNumbers)
+bool CSSParserFastPaths::isSimpleLengthPropertyID(CSSPropertyID propertyId, ValueRange& valueRange)
```
It would be a bit more idiomatic for this to return a std::optional. Maybe define a type like:

```
struct SimpleLengthProperties {
    ValueRange valueRange;
};

std::optional<SimpleLengthProperties> CSSParserFastPaths::simpleLengthPropertiesForPropertyID(CSSPropertyID propertyId) 
{
...
}
```

(probably not worth doing with this functional change though).

weinig commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -150,17 +150,35 @@ static inline bool parseSimpleAngle(std::span<const CharacterType> characters, C
 
-static RefPtr<CSSValue> parseSimpleLengthValue(CSSPropertyID propertyId, StringView string, CSSParserMode cssParserMode)
+template <typename CharacterType>
+static inline bool parseSimpleNumberOrPercentage(std::span<const CharacterType> characters, ValueRange valueRange, CSSUnitType& unit, double& number)
```
We don't have great documentation of these functions, but a comment indicating what differentiates "simple" from normal here would be quite useful for readers. 

(like above, changing to a std::optional return would be a bit more idiotic, but also like above, not worth doing with this functional change).

smfr replied:
Agree. Autogenerating more of this would be great.

weinig commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -441,10 +465,12 @@ static std::optional<SRGBA<uint8_t>> finishParsingHexColor(uint32_t value, unsig
 
-template<typename CharacterType> static std::optional<SRGBA<uint8_t>> parseHexColorInternal(std::span<const CharacterType> characters)
+template<typename CharacterType>
+static std::optional<SRGBA<uint8_t>> parseHexColorInternal(std::span<const CharacterType> characters)
```
@darinadler is pretty fond of having the `template<>` part all on the same line. 

I don't have strong opinions, but rather than ping pong these lines, we should probably standardize on a style for this.

smfr replied:
I prefer having `template<>` on its own line because it leave the next line looking mostly like a normal function.

darinadler replied:
We could debate this later. This is exactly the reason I like it on the same line! Because otherwise to me it looks too much like a non-template function.

weinig commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -950,17 +933,37 @@ static RefPtr<CSSValue> parseColorWithAuto(StringView string, const CSSParserCon
-    if (auto result = parseSimpleLengthValue(propertyID, string, context.mode))
-        return result;
-    if (propertyID == CSSPropertyDisplay)
+    switch (propertyID) {
```
I would love to find a way to automate this, at least a bit, from CSSProperties.json. (it memory serves, I did a bit of that already for keywords, but this would be a bit more).

We would have to interrogate the grammar of a property to see if one of our fast path code paths works for it (at the very least, generating isSimpleLengthPropertyID() should be quite feasible).

Not for this change, but something I think is worth noodling over.

smfr committed:
Add support for opacity to the CSS parser fast-path
https://bugs.webkit.org/show_bug.cgi?id=276646
rdar://131808782

Reviewed by Sam Weinig.

Add support for parsing opacity to CSSParserFastPaths, supporting numeric and percentage
forms.

Rearrange CSSParserFastPaths::maybeParseValue() to switch on the propertyID; previously we
we calling `parseSimpleTransform()` for things we knew were not a transform.

Hoist the call to `isSimpleLengthPropertyID()` out of `parseSimpleLengthValue()`. The
`isCSSViewportParsingEnabledForMode()` check there is obsolete and can probably be removed,
since we don't support `@viewport`. For simple length properties, we still fall through to
`parseKeywordValue()` for values like `auto`.

Support the `none` display value.

Remove `transformCanLikelyUseFastPath()` because it's faster to just call call
`parseSimpleTransformList()`.

* LayoutTests/fast/css/parsing-opacity-expected.txt:
* LayoutTests/fast/css/parsing-opacity.html:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isSimpleLengthPropertyID):
(WebCore::parseSimpleNumberOrPercentage):
(WebCore::parseSimpleLengthValue):
(WebCore::parseColorIntOrPercentage):
(WebCore::parseHexColorInternal):
(WebCore::parseNumericColor):
(WebCore::parseSimpleTransformList):
(WebCore::parseSimpleTransform):
(WebCore::parseDisplay):
(WebCore::parseOpacity):
(WebCore::CSSParserFastPaths::maybeParseValue):
(WebCore::transformCanLikelyUseFastPath): Deleted.
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
(WebCore::isValueOutOfRangeForProperty):

Canonical link: https://commits.webkit.org/281010@main
